ideas/289-perfect-forward-secrecy.md

```diff
@@ -0,0 +1,107 @@
+---
+id: 289-perfect-forward-secrecy
+title: Perfect Forward Secrecy
+status: draft
+created: 2018-07-23
+category: core
+lead-contributor: @cammellos
+contributors:
+    - @yenda
+    - @adamb
+exit-criteria: no
+success-metrics: no
+clear-roles: no
+future-iteration: no
+roles-needed:
+    - QA
+    - Clojure dev
+    - Go dev
+    - UX designer
+okrs:
+    - 
+    -
+---
+
+## Preamble
+
+    Idea: 289-perfect-forward-secrecy
+    Title: Perfect Forward Secrecy
+    Status: Draft
+    Created: 2018-07-23
+
+## Summary
+
+Pefect Forward Secrecy (PFS) is a must have for a secured messenging application. 
+The aim of this swarm is to implement PFS in Status and proceed to some architecture
+refactoring to make the communication protocol usable as a third party library.
+
+## Swarm Participants
+
+- Lead Contributor: @cammellos
+- Contributor: @yenda
+- Contributor: @adamb
+- QA:
+- UX:
+
+## Product Overview
+
+During Basel Meetup, PFS was defined as one of the key priorities.
+`Forward secrecy protects past sessions against future compromises of secret keys or passwords`
+This will require some changes and refactoring of the communication protocol.
+
+### User Stories
+
+As a user, I want Status to use PFS, so that my past communications are protected
+even if my secret keys are compromised (mitigation of potential information
+disclosure).
+
+As a third party developper, I want to be able to use the communication protocol
+with minimal setup and boilerplate, so that I can create bots or dapps that uses
+the protocol.
+
+### Requirements & Dependencies
+
+> Are there bugs or feature requests in other repositories that are part of this
+> Idea?
+
+### Security and Privacy Implications
+
+PFS is a critical functionnality for a secure messenger in order to reduce the
+risk of information disclosure. Our implementation should be audited and confirmed
+as a correct implementation of Signal Double Ratchet Algorithm
+(https://signal.org/docs/specifications/doubleratchet/) and X3DH Key Agreement 
```

Comment:
X3DH requires server, are we going to implement it?

Reply:
There's this discuss thread about how to avoid having a smart server: https://discuss.status.im/t/x3dh-and-one-time-key-avoid-smart-server-by-randomizing-the-one-time-key/276

Reply:
there's a more thorough description above, but basically, we are going to exchange the x3dh bundle through different channels, public chats / one to one / qr codes / contact codes / possibly ens and eventually swarm when ready, and will be refreshed periodically through contact updates. One time pre-key will only be used when the bundle is exchanged through a 1-to-1 message.